glean-core/rlb/src/private/labeled.rs

```diff
@@ -0,0 +1,389 @@
+    /// This also allows us to hide methods, that are only used internally
+    /// and should not be visible to users of the object implementing the
+    /// `Labeled<T>` trait.
+    pub trait ToLabeled {
+        /// The `glean_core` metric type representing the labeled metric.
+        type InnerCoreMetric: glean_core::metrics::MetricType + Clone;
+
+        /// Create a new metric object implementing this trait from the inner type.
+        fn from_inner(metric: Self::InnerCoreMetric) -> Self;
+
+        /// Create a new `glean_core` metric from the metadata.
+        fn new_inner(meta: crate::CommonMetricData) -> Self::InnerCoreMetric;
+    }
+
+    // `LabeledMetric<BooleanMetric>` is possible.
+    //
+    // See [Labeled Booleans](https://mozilla.github.io/glean/book/user/metrics/labeled_booleans.html).
+    impl ToLabeled for BooleanMetric {
+        type InnerCoreMetric = glean_core::metrics::BooleanMetric;
+
+        fn from_inner(metric: Self::InnerCoreMetric) -> Self {
+            BooleanMetric::new_from_meta(Arc::new(metric))
+        }
+
+        fn new_inner(meta: CommonMetricData) -> Self::InnerCoreMetric {
+            glean_core::metrics::BooleanMetric::new(meta)
+        }
+    }
+
+    // `LabeledMetric<StringMetric>` is possible.
+    //
+    // See [Labeled Strings](https://mozilla.github.io/glean/book/user/metrics/labeled_strings.html).
+    impl ToLabeled for StringMetric {
+        type InnerCoreMetric = glean_core::metrics::StringMetric;
+
+        fn from_inner(metric: Self::InnerCoreMetric) -> Self {
+            StringMetric::new_from_meta(Arc::new(metric))
+        }
+
+        fn new_inner(meta: CommonMetricData) -> Self::InnerCoreMetric {
+            glean_core::metrics::StringMetric::new(meta)
+        }
+    }
+
+    // `LabeledMetric<CounterMetric>` is possible.
+    //
+    // See [Labeled Counters](https://mozilla.github.io/glean/book/user/metrics/labeled_counters.html).
+    impl ToLabeled for CounterMetric {
+        type InnerCoreMetric = glean_core::metrics::CounterMetric;
+
+        fn from_inner(metric: Self::InnerCoreMetric) -> Self {
+            CounterMetric::new_from_meta(Arc::new(metric))
+        }
+
+        fn new_inner(meta: CommonMetricData) -> Self::InnerCoreMetric {
+            glean_core::metrics::CounterMetric::new(meta)
+        }
+    }
+}
+
+/// Marker trait for metrics that can be nested inside a labeled metric.
+///
+/// This trait is sealed and cannot be implemented for types outside this crate.
+pub trait AllowLabeled: private::ToLabeled {}
+
+// Implement the trait for everything we marked as allowed.
+impl<T> AllowLabeled for T where T: private::ToLabeled {}
+
+// We need to wrap the glean-core type: otherwise if we try to implement
+// the trait for the metric in `glean_core::metrics` we hit error[E0117]:
+// only traits defined in the current crate can be implemented for arbitrary
+// types.
+
+/// This implements the developer facing API for recording labeled metrics.
```

Comment:
```suggestion
/// This implements the specific API for recording labeled metrics.
```

or 

```suggestion
/// This implements the user-facing API for recording labeled metrics.
```

or at least

```suggestion
/// This implements the developer-facing API for recording labeled metrics.
```